Lib/pathlib.py

```diff
@@ -183,19 +183,18 @@ def resolve(self, path, strict=False):
             if strict:
                 return self._ext_to_normal(_getfinalpathname(s))
             else:
+                tail_parts = []  # End of the path after the first one not found
                 while True:
                     try:
                         s = self._ext_to_normal(_getfinalpathname(s))
                     except FileNotFoundError:
                         previous_s = s
-                        s = os.path.dirname(s)
+                        s, tail = os.path.split(s)
```

Comment:
Though for clarity, I'm fairly sure I introduced the call to `dirname` without knowing that we should avoid it. But it should be safer than `split` and `join` because I don't think those necessarily handle drive/machine names properly while `dirname` certainly should.